File: quantum/src/c/montecarlo.cc
#include "montecarlo.hh"
#include "config.h"
#include <sstream>
#include <iomanip>

/** @file montecarlo.cc
 * @brief Implementation of montecarlo.hh 
 */

/** Calculate Monte Carlo estimate with single level method */
void MonteCarloSingleLevel::evaluate(Statistics& stats) {
  std::shared_ptr<Path> x_path =
    std::make_shared<Path>(action->getM_lat(),
                           action->getT_final());

  // Burn-in phase
  for (unsigned int k=0;k<n_burnin;++k) {
    sampler->draw(x_path);
  }
  stats.reset();
  for (unsigned int k=0;k<n_samples;++k) {
    sampler->draw(x_path);
    /* Save (some) paths to disk? Edit file config.h */
#ifdef SAVE_PATHS
    if ( (SAVE_FIRST_PATH<=k) and (k<=SAVE_LAST_PATH) ) {
      std::stringstream filename;      
      filename << "path_";
      filename << std::setw(8) << std::setfill('0');
      filename << k << ".dat";
      x_path->save_to_disk(filename.str());
    }
#endif
    stats.record_sample(qoi->evaluate(x_path));
  }
}

/** Calculate mean and variance of difference in QoI */
void MonteCarloTwoLevel::evaluate_difference(Statistics& stats_fine,
                                             Statistics& stats_coarse,
                                             Statistics& stats_diff) {
  std::shared_ptr<Path> x_path =
    std::make_shared<Path>(fine_action->getM_lat(),
                           fine_action->getT_final());
  std::shared_ptr<Path> x_coarse_path =
    std::make_shared<Path>(coarse_action->getM_lat(),
                           coarse_action->getT_final());

  // Burn-in phase
  for (unsigned int k=0;k<n_burnin;++k) {
    coarse_sampler->draw(x_coarse_path);
    twolevel_step.draw(x_coarse_path,x_path);
  }

  // Sampling phase
  stats_diff.reset();
  for (unsigned int k=0;k<n_samples;++k) {
    coarse_sampler->draw(x_coarse_path);
    twolevel_step.draw(x_coarse_path,x_path);
    double qoi_fine = qoi->evaluate(x_path);
    double qoi_coarse = qoi->evaluate(x_coarse_path);
    stats_fine.record_sample(qoi_fine);
    stats_coarse.record_sample(qoi_coarse);
    stats_diff.record_sample(qoi_fine-qoi_coarse);
  }
}

File: quantum/src/python/montecarlo.py
import math
import numpy as np
from sampler import *

class MonteCarloSingleLevel(object):
    '''Single level MC Monte Carlo sampler
    
    :arg action: action to use
    :arg sample: Sampler object
    :arg qoi: Quantity of interest
    :arg n_samples: Number of samples for evaluation of QoI
    :arg n_burnin: Number of samples for burnin
    '''
    def __init__(self,action,
                 sampler,
                 qoi,
                 n_samples,
                 n_burnin):
        self._action = action
        self._sampler = sampler
        self._qoi = qoi
        self._n_samples = n_samples
        self._n_burnin = n_burnin

    def evaluate(self):
        '''Calculate the MCMC estimate (and its error) of the QoI
        by burning in and then evaluating the QoI on n_samples samples
        '''
        for i in range(self._n_burnin):
            self._sampler.draw()
        S = 0.0
        S_sq = 0.0
        for i in range(self._n_samples):
            x = self._sampler.draw()
            qoi_tmp = self._qoi(x)
            S += qoi_tmp
            S_sq += qoi_tmp**2

        mean = S/self._n_samples
        variance = 1./(self._n_samples*(self._n_samples-1))*(S_sq - self._n_samples*mean**2)
        return mean, math.sqrt(variance)

class MonteCarloTwoLevel(object):
    '''Two level MC Monte Carlo sampler
    
    :arg coarse_sampler: Exact sampler for coarse level
    :arg coarse_action: Coarse level action
    :arg action: action to use
    :arg sample: Sampler object
    :arg qoi: Quantity of interest
    :arg n_samples: Number of samples for evaluation of QoI
    :arg n_burnin: Number of samples for burnin
    '''
    def __init__(self,
                 coarse_sampler,
                 coarse_action,
                 action,
                 qoi,
                 n_samples,
                 n_burnin):
        self._qoi = qoi
        self._n_samples = n_samples
        self._n_burnin = n_burnin
        self._twolevel_sampler = TwoLevelMetropolisSampler(coarse_sampler,
                                                           coarse_action,
                                                           action,
                                                           verbosity=0)

    def evaluate(self):
        '''Calculate the MCMC estimate (and its error) of the QoI
        by burning in and then evaluating the QoI on n_samples samples
        '''
        for i in range(self._n_burnin):
            x_coarse, x_fine = self._twolevel_sampler.draw()
        S = 0.0
        S_sq = 0.0
        for i in range(self._n_samples):
            x_coarse, x_fine = self._twolevel_sampler.draw()
            qoi_tmp = self._qoi(x_fine)
            S += qoi_tmp
            S_sq += qoi_tmp**2

        mean = S/self._n_samples
        variance = 1./(self._n_samples*(self._n_samples-1))*(S_sq - self._n_samples*mean**2)
        return mean, math.sqrt(variance)

    def evaluate_difference(self):
        '''Calculate the difference between the QoI on the fine- and coarse
        level and return this difference and its variance.
        '''
        for i in range(self._n_burnin):
            x_coarse, x_fine = self._twolevel_sampler.draw()
        S = 0.0
        S_sq = 0.0
        for i in range(self._n_samples):
            x_coarse, x_fine = self._twolevel_sampler.draw()
            qoi_tmp = self._qoi(x_fine)-self._qoi(x_coarse)
            S += qoi_tmp
            S_sq += qoi_tmp**2

        mean = S/self._n_samples
        variance = 1./(self._n_samples-1)*(S_sq - self._n_samples*mean**2)
        return mean, variance
